src/kontrol/prove.py

```diff
@@ -494,6 +513,139 @@ def _init_cterm(
+    *,
+    setup_cterm: CTerm | None = None,
+    calldata: KInner | None = None,
+    callvalue: KInner | None = None,
+) -> CTerm:
+    contract = KEVM.account_cell(
+        KVariable('CONTRACT'),
+        KVariable('CONTRACT_BAL'),
+        program,
+        KVariable('CONTRACT_STORAGE'),
+        KVariable('CONTRACT_STORAGE'),
+        KVariable('CONTRACT_NONCE'),
+    )
+    acct_from = KEVM.account_cell(
+        KVariable('ACCT_FROM'),
+        KVariable('ACCT_FROM_BAL'),
+        KVariable('ACCT_FROM_CODE'),
+        KVariable('ACCT_FROM_STORAGE'),
+        KVariable('ACCT_FROM_ORIG_STORAGE'),
+        KVariable('ACCT_FROM_NONCE'),
+    )
+    schedule = KApply('SHANGHAI_EVM')
+
+    valid_callvalue = callvalue if callvalue is not None else KVariable('CALLVALUE')
+
+    init_subst = {
+        'MODE_CELL': KApply('NORMAL'),
+        'SCHEDULE_CELL': schedule,
+        'ID_CELL': KVariable('ACCT_FROM'),
+        'ACTIVE_CELL': FALSE,
+        'STATIC_CELL': FALSE,
+        'GAS_CELL': KEVM.inf_gas(KVariable('VGAS')),
+        'LOCALMEM_CELL': KVariable('LM', sort=KSort('Bytes')),
+        'K_CELL': KApply(
+            '_________EVM_InternalOp_CallOp_Int_Int_Int_Int_Int_Int_Int',
+            [
+                KApply('CALL_EVM_CallOp'),
+                intToken(1),
+                KVariable('CONTRACT'),
+                valid_callvalue,
+                KVariable('ARGSTART'),
+                intToken(method.argwidth),
+                KVariable('RETSTART'),
+                intToken(method.retwidth),
+            ],
+        ),
+        'ACCOUNTS_CELL': KEVM.accounts(
+            [
+                contract,  # test contract address
+                acct_from,
+                Foundry.account_CHEATCODE_ADDRESS(KApply('.Map')),
+            ]
+        ),
+        'ISREVERTEXPECTED_CELL': FALSE,
+        'ISOPCODEEXPECTED_CELL': FALSE,
+        'ISEVENTEXPECTED_CELL': FALSE,
+        'ISCALLWHITELISTACTIVE_CELL': FALSE,
+        'ISSTORAGEWHITELISTACTIVE_CELL': FALSE,
+    }
+
+    constraints = [
+        mlEqualsTrue(KApply('_>=Int_', KVariable('ACCT_FROM_BAL'), intToken(0))),
+        mlEqualsTrue(KApply('#rangeNonce(_)_WORD_Bool_Int', KVariable('ACCT_FROM_NONCE'))),
+        mlEqualsTrue(
+            notBool(KApply('#isPrecompiledAccount(_,_)_EVM_Bool_Int_Schedule', [KVariable('CONTRACT'), schedule]))
+        ),
+        mlEqualsTrue(ltInt(KVariable('CALLDEPTH_CELL'), intToken(1024))),
+        mlEqualsTrue(
+            KApply(
+                '_==Int_',
+                [
+                    KApply(
+                        '#asWord(_)_EVM-TYPES_Int_Bytes',
+                        KApply(
+                            '#range(_,_,_)_EVM-TYPES_Bytes_Bytes_Int_Int',
+                            [KVariable('LM'), KVariable('ARGSTART'), intToken(4)],
+                        ),
+                    ),
+                    KEVM.abi_selector(method.signature),
+                ],
+            )
+        ),
+    ]
```

Comment:
Should these constraints be added for _any_ proof? They seem prettygeneric. Maybe it can be added to `KEVM.add_constraints` instead?

Reply:
This is where that `KEVM.add_invariant` method is, which does exactly that: https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kevm.py#L241